go/vt/vttablet/tabletmanager/rpc_replication.go

```diff
@@ -419,8 +420,30 @@ func (tm *TabletManager) demotePrimary(ctx context.Context, revertPartialFailure
 		if err := tm.QueryServiceControl.SetServingType(tablet.Type, logutil.ProtoToTime(tablet.PrimaryTermStartTime), false, "demotion in progress"); err != nil {
 			return nil, vterrors.Wrap(err, "SetServingType(serving=false) failed")
 		}
+
+		// Note that due to an upstream semi-sync bug the semi-sync ACK request may be lost, causing the
+		// current primary to wait forever for the ACK. See:
+		//   https://bugs.mysql.com/bug.php?id=104015
+		//   https://bugs.mysql.com/bug.php?id=101056
+		// This in turns causes the statement to disable [Super]ReadOnly mode to block on a mutex that
+		// the semi-sync plugin is holding while waiting for the ACK. This then can cause the shard to
+		// be left in an unhealthy state without a serving PRIMARY. So here we need to disable SemiSync
+		// which will ensure that we are unblocked by this scenario (tablet repair should also fix this
+		// as needed via fixSemiSync()).
+		if wasSemiSyncSourceEnabled {
+			log.Infof("DemotePrimary disabling source semi-sync")
+			if err := tm.MysqlDaemon.SetSemiSyncEnabled(false, wasSemiSyncReplicaEnabled); err != nil {
```

Comment:
It is safe to do this because we have already shut down the tablet server by calling `SetServingType(..., serving=false, ...)`. Tablet Server shutdown gives a grace period to open transactions and then force-kills them. Even if a grace period is not set, a transaction that is hanging because of the semi-sync bug will get killed after the transaction timeout.
Nice work @mattlord